src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -138,20 +141,26 @@ void ShenandoahOldHeuristics::slide_pinned_regions_to_front() {
     }
   }
 
+  // If we could not find an unpinned region, it means there are no slots available
+  // to move up the pinned regions. In this case, we just reset our next index in the
+  // hopes that some of these regions will become unpinned before the next mixed
+  // collection. We may want to bailout of here instead, as it should be quite
+  // rare to have so many pinned regions and may indicate something is wrong.
   if (write_index == NOT_FOUND) {
-    if (_first_pinned_candidate > 0) {
-      _next_old_collection_candidate = _first_pinned_candidate;
-    }
+    assert(_first_pinned_candidate != NOT_FOUND, "Should only be here if there are pinned regions.");
+    _next_old_collection_candidate = _first_pinned_candidate;
     return;
   }
 
   // Find pinned regions to the left and move their pointer into a slot
-  // that was pointing at a region that has been added to the cset.
-  // [ r p r p p p r ]
-  //       ^
-  //       | Write pointer is here. We know this region is already in the cset
-  //       | so we can clobber it with the next pinned region we find.
-  for (size_t search = write_index - 1; search > _first_pinned_candidate; --search) {
+  // that was pointing at a region that has been added to the cset. We
+  // are done when the leftmost pinned region has been slid up.
+  // [ r p r x p p p r ]
+  //         ^       ^
+  //         |       | next region for mixed collections
+  //         | Write pointer is here. We know this region is already in the cset
+  //         | so we can clobber it with the next pinned region we find.
+  for (int32_t search = write_index - 1; search >= (int32_t)_first_pinned_candidate; --search) {
     RegionData& skipped = _region_data[search];
     if (skipped._region->is_pinned()) {
       RegionData& added_to_cset = _region_data[write_index];
```

Comment:
Yes, that's a good point. I shall also remove the assert for the same reason (it may encounter a slot for a pinned region that has been moved up and such a region should _not_ be in the cset).